Openai back

LucileTech opened the pull request:
## I'm submitting a ...

- [ ] bug report
- [x] feature request

## What is the current behavior?
post endpoint for openai

## What is the expected behavior?
return a response to the user message
curl -X POST "http://127.0.0.1:8000/api/chat" -H "Content-Type: application/json" -d '{"message": "Hello, how are you?"}'

## What is the motivation / use case for changing the behavior?
New feature

huyfififi commented:
You might want to run `pdm add openai` in the backend directory to update the dependencies.
I would appreciate it if you could use the docker compose I have added to replicate the environment. I'm not sure if you are familiar with that technology, but if not, this is a good opportunity to learn Docker Compose, which is used by many teams. Please see backend/README.md for its usage

huyfififi commented:
and in general, a pull request in progress should be marked as a draft, indicating that the pull request is not ready for reviews. Once it's ready, please assign reviewers, so that they will get email notifications and be aware of your pull request

huyfififi commented on backend/src/backend/app.py:
```diff
@@ -115,6 +128,34 @@ async def get_tasks(user_id: Annotated[str, Depends(get_user_id)]):
+
+@app.post("/api/chat")
+async def chat(request: Request):
+    data = await request.json()
```
In FastAPI, request body is defined to validate and enforce its structure. 
https://fastapi.tiangolo.com/tutorial/body/

How about we create a Pydantic model such as

```
class Chat(BaseModel):
  message: str
```

not sure about the requirements of this API endpoint tho

huyfififi replied:
by doing this, we don't need to write the validation at line 139

huyfififi commented on backend/src/backend/app.py:
```diff
@@ -115,6 +128,34 @@ async def get_tasks(user_id: Annotated[str, Depends(get_user_id)]):
+# Open AI chat
+openai.api_key = OPENAI_API_KEY 
+
+@app.post("/api/chat")
```
In FastAPI, specifying `response_model` is a common practice

LucileTech commented:
> You might want to run `pdm add openai` in the backend directory to update the dependencies. (We use `PDM` to manage Python dependencies) I would appreciate it if you could use the docker compose I have added to replicate the environment. I'm not sure if you are familiar with that technology, but if not, this is a good opportunity to learn Docker Compose, which is used by many teams. Please see backend/README.md for its usage

Thank you so much for your review [huyfififi](https://github.com/huyfififi) !
For this part, I did run Docker and added OpenAI to the dependencies. I am curious why you thought I did not. Did I miss adding something to the GitHub repo?

huyfififi commented:
@LucileTech Thanks for the quick fix on the FastAPI practices 👍 

for this part

> Thank you so much for your review [huyfififi](https://github.com/huyfififi) ! For this part, I did run Docker and added OpenAI to the dependencies. I am curious why you thought I did not. Did I miss adding something to the GitHub repo?

We use `PDM`, which is a Python package manager, and it generates `pdm.lock` file every time we update dependencies by `pdm add {package name}`.
It works in a similar way as `package-lock.json` in the frontend; the lock file is used to freeze not only the top-level packages we specify in `pyproject.toml` but also all the sub dependencies.

When using a package manager such as `PDM`, updating `pyproject.toml` directly is not the right way, corrupting the lock file. The lock file is used in Docker Compose and GitHub Actions, so you can see the error in the Actions tab

huyfififi commented:
Are you interested in the backend development? It would be super helpful if you could add a test to get validation errors when sending malformed request payload.

testing OpenAI Integration (or any kind of integrations) is very difficult, so for now, we don't need to test that case.

```
response = client.post(json={"test": "test"})
assert resposne.status_code == 400
assert response.json() == {"error_message": ""}
```

something like this would be enough for now.

If you are not interested in doing that, pls feel free to ignore this comment. I'll add sample tests so that you can refer to later

huyfififi commented on backend/src/backend/app.py:
```diff
@@ -114,6 +128,39 @@ async def get_tasks(user_id: Annotated[str, Depends(get_user_id)]):
+
+# Pydantic models for request and response
+class ChatRequest(BaseModel):
+    message: str
```
```
message: str = Field(..., min_length=1)
```

if you want to block empty strings

huyfififi commented on backend/src/backend/app.py:
```diff
@@ -114,6 +128,39 @@ async def get_tasks(user_id: Annotated[str, Depends(get_user_id)]):
+async def chat(chat_request: ChatRequest):
+    user_message = chat_request.message
+
+    if not user_message:
```
I think this validation is already done on the FastAPI side, and we don't need to write it.
Writing tests might provide how this API endpoint works. Having tests also helps reviewers understand the logic by providing input/output examples.

Please feel free to reach out to me if you need some help.

huyfififi commented on backend/src/backend/app.py:
```diff
@@ -114,6 +128,39 @@ async def get_tasks(user_id: Annotated[str, Depends(get_user_id)]):
+        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="No message provided")
+
+    # Prepend instruction to prompt for a brief response
+    prompt = f"Please respond in a brief phrase: {user_message}"
```
shouldn't we concatenate user_message to this string?

huyfififi commented:
`nit` please rebase with `main`. there are too many workflows triggered for this pull request

LucileTech commented:
> Are you interested in the backend development? It would be super helpful if you could add a test to get validation errors when sending malformed request payload.
> 
> testing OpenAI Integration (or any kind of integrations) is very difficult, so for now, we don't need to test that case.
> 
> ```
> response = client.post(json={"test": "test"})
> assert resposne.status_code == 400
> assert response.json() == {"error_message": ""}
> ```
> 
> something like this would be enough for now. (see `backend/src/backend/tests.py`)
> 
> If you are not interested in doing that, pls feel free to ignore this comment. I'll add sample tests so that you can refer to later

Yes, I would like to try! Could I use 422 status code or do you want me to customize to 400?
As I understood, FastAPI responds with a 422 Unprocessable Entity status code for malformed request payloads, but I can adjust the endpoint to return a 400 Bad Request

huyfififi commented:
Yes, your understanding is correct! FastAPI returns 422 as the default, but IMO, 400 Bad Request is more appropriate and popular. It clearly indicates that there is an error in the request payload.

In other popular frameworks such as Django REST framework, 400 Bad Request will be retuned when request payload has problems.

You might want to try this but please confirm
```
# override the default status for malformed request payload 422 -> 400
@app.exception_handler(RequestValidationError)
async def validation_exception_handler(request, exc):
    return PlainTextResponse(str(exc), status_code=400)
```

huyfififi commented on backend/src/backend/app.py:
```diff
@@ -114,6 +128,36 @@ async def get_tasks(user_id: Annotated[str, Depends(get_user_id)]):
+openai.api_key = OPENAI_API_KEY 
+
+# Pydantic models for request and response
+class ChatRequest(BaseModel):
```
Python has an official guideline for code styling, called PEP8, and it recommends to have two blank lines between top-level functions/classes

https://peps.python.org/pep-0008/#blank-lines
> Surround top-level function and class definitions with two blank lines.

huyfififi replied:
not sure why `ruff` did not report this... maybe I made mistakes configuring the GitHub Actions.
From my experience, many teams request Python engineers to run `flake8` and `black` (code formatter) before committing changes, just for your information

LucileTech replied:
> run `flake8` and `black`

Thank you for letting me know, I will configure flake8 and black!

huyfififi commented on backend/src/backend/tests.py:
```diff
@@ -40,3 +40,13 @@ def test_list_tasks_unauthorized_user():
+
+# OpenAI
+
+def test_chat_endpoint_malformed_request():
```
👍

huyfififi commented:
can you also add a test with an empty input? `{"message": ""}`

huyfififi commented on backend/src/backend/tests.py:
```diff
@@ -40,3 +40,11 @@ def test_list_tasks_unauthorized_user():
+    response = client.post("/api/chat", json={"test": "test"})
+    assert response.status_code == 400
+    assert response.json() == {"error_message": "Malformed request"}
+
```
please remove redundant spaces

huyfififi commented on backend/src/backend/app.py:
```diff
@@ -115,6 +131,49 @@ async def get_tasks(user_id: Annotated[str, Depends(get_user_id)]):
+
+
+# override the default status for malformed request payload 422 -> 400
+@app.exception_handler(RequestValidationError)
```
can we move this and cors middleware settings right after `app.add_middleware`? relevant code (FastAPI configuration this case `app`) should be gathered to one place.

huyfififi commented:
almost done!

huyfififi commented on backend/src/backend/app.py:
```diff
@@ -115,6 +131,49 @@ async def get_tasks(user_id: Annotated[str, Depends(get_user_id)]):
+async def validation_exception_handler(request: Request, exc: RequestValidationError):
+    return JSONResponse(
+        status_code=status.HTTP_400_BAD_REQUEST,
+        content={"error_message": "Malformed request"}
```
I think we should follow this https://fastapi.tiangolo.com/tutorial/handling-errors/#override-request-validation-exceptions

```
@app.exception_handler(RequestValidationError)
async def validation_exception_handler(request, exc):
    return PlainTextResponse(str(exc), status_code=400)
```

With this approach, the FastAPI (Pydantic) validation error will provide more detailed error messages.
I don't feel returning "Malformed request" without details for any type of errors is a good approach

huyfififi replied:
I know you're still working on this pull request, but just in case

huyfififi commented on backend/src/backend/tests.py:
```diff
@@ -40,3 +40,22 @@ def test_list_tasks_unauthorized_user():
+
+
+# Chat enpoint
+# Test for malformed request
```
The test function names already contain the information, so these comments (lines 46 and 55) are redundant.

huyfififi commented on backend/src/backend/tests.py:
```diff
@@ -40,3 +40,22 @@ def test_list_tasks_unauthorized_user():
+    assert response.status_code == 400
+    assert "type" in response.text
+    assert "loc" in response.text
+    assert "msg" in response.text
```
Please add more specific assertions, especially `msg` because it's the most important information clients get.

```
assert response.json() == {
  "detail": [
    {
    }
  ]
}
```

huyfififi commented:
@LucileTech 
It's been a while since this pull request was created, and I would like to close this. Taking too much time to merge a pull request makes it challenging to keep it up to date with the default branch, and the chances of getting conflicts are high.
Are there any improvements you are about to make?

LucileTech commented:
> @LucileTech It's been a while since this pull request was created, and I would like to close this. Taking too much time to merge a pull request makes it challenging to keep it up to date with the default branch, and the chances of getting conflicts are high. Are there any improvements you are about to make?

Of course, we can merge! Thank you very much for your help and insights; they were very helpful and instructive. I can improve the tests in another PR or push a commit here before the merge.

huyfififi reviewed (approved):
Looks good to me 👍 
Let's improve the tests in another PR
